### ISSUE TITLE
Keyboard Accessibility

### ISSUE THREAD

**Issue by mrmurphy** [mbylstra/atom-elm-navigator#7]
Wow, what a cool idea! I love aesthetic and the functional design 👍 

It'd be super fantastic if the navigator could be hidden by default, and then a command offered which would expose the navigator and focus the filter field. After typing in the filter field, if the user were to use the up/down arrow keys, results would be highlighted in the navigator. Pressing enter with a result highlighted would navigate to that result and close the navigator.

That'd really help those of us who spend most of our time on the keyboard.

Again, thanks for the time and effort!

**Comment by mbylstra** [comment]
Thanks Murphy!

Very good points about keyboard accessibility - it would be great to not have to leave the keyboard. That's a habit I want to get into more. I'm going to give this a high priority. 

I'm going to split "hidden by default" into a separate issue though.

**Comment by DavidLGoldberg** [comment]
Yeah, I recommend taking inspiration from cntrl + 0, cmd + tab (tree view) along with vim style j/k to scroll up and down :)

Also, I'm working on a rewrite of my package Jumpy...ultimately to get it to jump files in the tree view, but making the architecture more generic, so I should be able to add entries in the navigator as well.  If only I could learn elm faster so I get the rewrite done :)

Also, have you given thought to making this thing not just for elm......there are / were several attempts at something similar but I don't know of any good ones that stuck, and if they did I doubt they're generic, and there *really* should be one.

I guess a good measure of flexibility for the api would be, could most of your elm specific stuff stay in elm and register for presentation in JS via generic navigator api calls

**Comment by DavidLGoldberg** [comment]
I just remembered about "symbols" cmd+r (I don't use it much in Atom).

I did a little research and I'll share it, in case any is new to you:
The built in atom uses "ctags" particularly an old "exuberant ctags" as opposed to the newer "universal ctags" (that is still more actively maintained)  

Here is the atom package btw for symbols (built in):
https://atom.io/packages/symbols-view

symbol competitors ranked by downloads:
https://atom.io/packages/symbols-tree-view
https://atom.io/packages/symbols-list
https://atom.io/packages/symbols-view-plus (only 1100 downloads but looks real promising)

The last section in the language doc talks about this a bit:
https://atom.io/packages/language-elm
but that strategy from what I can tell wouldn't help you here.

Here's another package that does elm ctags, although it doesn't look as involved as your regex from what I can tell, and I'm not sure how easy it is to add to atom.
https://github.com/kbsymanz/ctags-elm

**Comment by mbylstra** [comment]
In an earlier version of this I used ctags, but I ended up writing my own parser/regex. My memory is a little hazy, but these were the main issues (or something like this):

- I remember there was some extra things you needed to set up as a user to get it to work, so it was annoying from an installation point of view.
- Ideally I'd want the nav bar to show a real-time, or as close to real-time view of the page as possible. I remember only being able to set it up to update on save, so things would quickly get out of date. With a custom implementation, I was able create an algorithm that was wait until the user had idled for a couple of seconds, and update it then - so it's quite close to real time. I think this is similar to how Jetbrains products do expensive updates. It works quite well!  
- Despite being written in C, it was actually quite slow (perhaps because it was searching many more files and directories than necessary?)

I decided to make this an Elm-only plugin, because frankly I mostly program in Elm these days, so I wasn't interested in trying support multiple languages. Because it's Elm specific I can add Elm specific features such as highlighting the **update**, **view** and **init** functions for quick access. I've found this to be one of the most useful features. I have other Elm specific ideas that I haven't got round to yet.

I'm hoping to find some time to do some more work on this plugin soon.
